File: Algorithms/Implementation/Encryption/Encryption.java
// Author: Atharv Damle
// To encrypt an input using a matrix. Size of matrix = square root of the size of input.
// Full Question: https://www.hackerrank.com/challenges/encryption/problem

import java.io.*;
import java.util.*;

public class Solution {

    static String encryption(String s) {
        s = s.replaceAll("\\s", "");
        
        int l = s.length();
        // x represents the number of rows.
        int x = (int) Math.sqrt(l);
        
        // The matrix
        char a[][];
        
        // Number of columns in matrix.
        int c = 0;
        
        // If the size is a perfect square...
        if (x * x == l) 
        {
            a = new char[x][x];
            c = x;
        }
        else 
        {
            c = x + 1;
            // If the input does not fit in
            // a matrix of size x * c...
            if (c * x < l) {
                x++;
            }
            a = new char[x][c];
        }

        // Creating the transpose of the matrix without 
        // storing the input in the original order by
        // inputting columnwise (ie vertically into the matrix)
        int p = 0;
        int q = 0;
        for (int i = 0; i < l; i++) 
        {
            a[p][q] = s.charAt(i);
            q++;
            
            if (q == c) 
            {
                q = 0;
                p++;
            }
        }
        
        // Make the matrix into a string with
        // equal size words.
        s = "";
        for (int i = 0; i < c; i++) 
        {
            for (int j = 0; j < x; j++) 
            {
                // In case any empty element has a garbage value...
                if (Character.isLetter(a[j][i]))
                {
                    s += a[j][i];
                }
            }
            s += " ";
        }

        return s;
    }
    
    public static void main(String[] args) throws IOException 
    {
        BufferedWriter b = new BufferedWriter(new FileWriter(System.getenv("OUTPUT_PATH")));
        Scanner sc = new Scanner(System.in);
        String s = sc.nextLine();
        
        // Measuring time.
        long t = System.nanoTime();
        
        // Calling the function
        String result = encryption(s);
        
        // Elapsed time...
        System.out.println("Elapsed: " + (System.nanoTime() - t) + "ns");// * Math.pow(10, -9) + "s");
        
        b.write(result);
        b.newLine();

        b.close();

        sc.close();
    }
}

File: Algorithms/Implementation/Encryption/Encryption.c
// Author: Atharv Damle
// To encrypt an input using a matrix. Size of matrix = square root of the size of input.
// Full Question: https://www.hackerrank.com/challenges/encryption/problem

#include <math.h>
#include <stdbool.h>
#include <stdio.h>
#include <stdlib.h>
#include <ctype.h>
#include <time.h>

// Macros to measure time.
#define TICK(X) clock_t X = clock()
#define TOCK(X)                                                                \
  printf("elapsed time %s: %g sec.\n", (#X),                                   \
         (double)(clock() - (X)) / CLOCKS_PER_SEC)

int min(int a, int b)
{
    if (a > b)
    {
        return a;
    }
    return b;
}

char* encryption(char* s) {

  // Remove spaces from string
  int l = strlen(s);
  int p = 0;
  for (int i = 0; i < l; i++)
  {
      if (s[i] != ' ')
      {
          s[p] = s[i];
          p++;
      }
  }
  s[p] = '\0';
  l = p;

  // x must be initialized.
  int x = 0;
  x = (int)sqrt(l);

  // Decide the size of the matrix. x = number of rows
  // c = number of columns
  int c = 0;
  if (x * x == l) 
  {
    c = x;
  } 
  else 
  {
    c = x + 1;
    if (c * x < l) 
    {
      x++;
    }
  }
  
  // Create the matrix.
  char a[x][c];
  int m = min(p + c, l);
  // including the spaces.
  s = realloc(s, m);

  // Add the string into the matrix.
  p = 0;
  int q = 0;
  for (int i = 0; i < l; i++) {
    a[p][q] = s[i];
    q++;

    if (q == c) {
      // System.out.println(a[p][q]);
      q = 0;
      p++;
    }
  }

  // Create new string from the matrix (with equally sized words)
  p = 0;
  for (int i = 0; i < c; i++) {
    for (int j = 0; j < x; j++) {
      if (isalpha(a[j][i])) {
        s[p] = a[j][i];
        p++;
      }
    }
    s[p] = ' ';
    p++;
  }
  s[m - 1] = '\0';

  return s;
}

int main()
{
    // Open file
    FILE* fptr = fopen(getenv("OUTPUT_PATH"), "w");

    // Accept input
    
    // Optional - Start the clock (To measure speed)
    TICK(TIME_A);
    
    // Call the function
    char* result = encryption(s);
    
    // If clock was started, stop the clock and print the elapsed time to stdout
    //TOCK(TIME_A);

    // Write output of the function to the file
    fprintf(fptr, "%s\n", result);

    fclose(fptr);

    return 0;
}
